proposals/010-authorizer.md

```diff
@@ -0,0 +1,212 @@
+It will also use the same implied operation semantics as implemented by Kafka itself, such as where `ALTER` implies `DESCRIBE`, as described by 
+`org.apache.kafka.common.acl.AclOperation`.
+
+There is one deviation.  The filter will implement a `CONNECT` authorization check on the `CLUSTER` early, as the connection is made, once the principal is known.  This allows the Authorizer filter to be used to gate access to virtual clusters.
+* In the case of SASL, this will be performed on receipt of the Sasl Authentication Response.  If the authorization check fails, the authentication will fail with an authorization failure and the connection closed.
+* In case of TLS client-auth, this will be performed on receipt of the first request message.   If the authorization check fails, a short circuit response will be sent containing an authorization failure and the connection closed.  This feature won’t be part of the initial scope.
+
+The filter will support messages that express topic identity using topic ids (i.e. those building on [KIP-516](https://cwiki.apache.org/confluence/display/KAFKA/KIP-516%3A+Topic+Identifiers)).  It will resolve the topic id into a topic name before making the authorization check.
+
+### File based Authorizer implementation
+
+The initial scope will include an Authorizer implementation that is backed by authorization rules expressed in a separate file.  This file will associate principals with ACL rules capable of expressing an allow-list of resources.
+The initial version will be restricted to expressing allow lists of topics, but future version will extend this to allow for rules to be express about other resource types.
+
+### APIs
+
+#### Authorizer Filter
+
+Filter Configuration:
+
+```yaml
+type: AuthorizerFilter
+config:
+  authorizer: FileBasedAllowListAuthorizer
+  authorizerConfig:
+    rulesFile: /path/to/allow-list.yaml
+```
+
+Java APIs:
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.Authorizer, except is asynchronous in nature.
+interface Authorizer {
+    CompletionStage<List<AuthenticationResult>> authorize(AuthorizableRequestContext context, List<Action> actions);
+}
+```
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.AuthorizableRequestContext
+interface AuthorizableRequestContext {
+    String principal();
+    // scope for methods such as requestType(), requestVersion()  etc to be added in future.
+}
+```
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.Action
+record Action(
+    AclOperation aclOperation,
+    ResourcePattern resourcePattern) {
+}
+```
+
+```java
+// The following types are inspire by the Kafka classes of the same name and have the same role.  However, interfaces are used
+// rather than enums to allow for extensibility (using pattern suggested by https://www.baeldung.com/java-extending-enums)
+interface AclOperation {
+    String operationName();
+}
+
+enum CoreAclOperation implements AclOperation {
```

Comment:
One thing I don't like about this is that there's no safety mechanism to ensure that the operation is compatible with the resource type.

I was playing with a pattern last week where we would represent a resource type simply as an `enum` of the operations it supports (your `ResourceType` becomes a `Class` of that `enum`). Thus `TopicResource` ends up with `READ`, `WRITE`, `DESCRIBE` etc and ClusterResource ends up with `CONNECT`, `DESCRIBE` etc, and you can construct a type safe API where asking for an authorization decision on `TopicResource.CONNECT ` is a compile time error.